git/index.go

```diff
@@ -221,14 +222,44 @@ const (
 //
 // As a special case, if something is added as Stage0, then Stage1-3 entries
 // will be removed.
-func (g *Index) AddStage(c *Client, path IndexPath, mode EntryMode, s Sha1, stage Stage, size uint32, mtime int64, createEntry bool) error {
+func (g *Index) AddStage(c *Client, path IndexPath, mode EntryMode, s Sha1, stage Stage, size uint32, mtime int64, opts UpdateIndexOptions) error {
 	if stage == Stage0 {
 		defer g.RemoveUnmergedStages(c, path)
 	}
 
+	replaceEntriesCheck := func() error {
+		// If replace is true then we search for any entries that
+		//  should be replaced with this one.
+		newObjects := make([]*IndexEntry, 0, len(g.Objects))
+		for _, e := range g.Objects {
+			if strings.HasPrefix(string(e.PathName), string(path)+"/") {
+				if !opts.Replace {
+					return fmt.Errorf("There is an existing file %s under %s, should it be replaced?", e.PathName, path)
+				}
+				continue
+			} else if strings.HasPrefix(string(path), string(e.PathName)+"/") {
+				if !opts.Replace {
+					return fmt.Errorf("There is a parent file %s above %s, should it be replaced?", e.PathName, path)
+				}
+				continue
+			}
+
+			newObjects = append(newObjects, e)
+		}
+
+		g.Objects = newObjects
+
+		return nil
+	}
+
 	// Update the existing stage, if it exists.
 	for _, entry := range g.Objects {
 		if entry.PathName == path && entry.Stage() == stage {
+			err := replaceEntriesCheck()
+			if err != nil {
```

Comment:
if you replace this (and the other one) with `if err := replaceEntriesCheck(); err != nil`, I'll merge it.